### ISSUE TITLE
metasploit error after kali update

### ISSUE THREAD

**Issue by surfer663** [rapid7/metasploit-framework#14673]
## Steps to reproduce

How'd you do it?

1. ...msfconsole>
2. search shodan>use auxiliary/gather/shodan_search>set shodan_apikey 000000000000(example)
3. ...set query webcam>run

This section should also tell us any relevant information about the
environment; for example, if an exploit that used to work is failing,
tell us the victim operating system and service versions.

## Were you following a specific guide/tutorial or reading documentation?
no

If yes link the guide/tutorial or documentation you were following here, otherwise you may omit this section.

## Expected behavior should have listed all ip addresses with countries

What should happen?got errors

## Current behavior errors

What happens instead? errors : [-] Auxiliary failed: JSON::ParserError 809: unexpected token at '<html>
<head><title>400 Bad Request</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
<hr><center>cloudflare</center>
</body>
</html>
'
[-] Call stack:
[-]   /usr/share/metasploit-framework/vendor/bundle/ruby/2.7.0/gems/json-2.5.1/lib/json/common.rb:216:in `parse'
[-]   /usr/share/metasploit-framework/vendor/bundle/ruby/2.7.0/gems/json-2.5.1/lib/json/common.rb:216:in `parse'
[-]   /usr/share/metasploit-framework/vendor/bundle/ruby/2.7.0/gems/activesupport-5.2.4.4/lib/active_support/json/decoding.rb:23:in `decode'
[-]   /usr/share/metasploit-framework/modules/auxiliary/gather/shodan_search.rb:72:in `shodan_query'
[-]   /usr/share/metasploit-framework/modules/auxiliary/gather/shodan_search.rb:118:in `run'
[*] Auxiliary module execution completed

### Metasploit version  metasploit v6.0.28-dev

Get this with the `version` command in msfconsole (or `git log -1 --pretty=oneline` for a source install).

## Additional Information
If your version is less than `5.0.96`, please update to the latest version and ensure your issue is still present.

If the issue is encountered within `msfconsole`, please run the `debug` command using the instructions below. If the issue is encountered outisde `msfconsole`, or the issue causes `msfconsole` to crash on startup, please delete this section.

1. Start `msfconsole`
2. Run the command `set loglevel 3`
3. Take the steps necessary recreate your issue
4. Run the `debug` command
5. Copy all the output below the `===8<=== CUT AND PASTE 
6.EVERYTHING BELOW THIS LINE ===8<===` line and make sure to **REMOVE ANY SENSITIVE INFORMATION.**

##  Module/Datastore

The following global/module datastore, and database setup was configured before the issue occurred:
<details>
<summary>Collapse</summary>

```
[framework/core]
loglevel=3
```

</details>

##  History

The following commands were ran during the session and before this issue occurred:
<details>
<summary>Collapse</summary>

```
1183   set loglevel 3
1184   debug
```

</details>

##  Framework Errors

The following framework errors occurred before the issue occurred:
<details>
<summary>Collapse</summary>

```
[01/28/2021 12:50:27] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/28/2021 12:55:27] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:55:27] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:55:27] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/28/2021 12:55:27] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/28/2021 12:59:18] [e(0)] core: Auxiliary failed - JSON::ParserError 809: unexpected token at '<html>
<head><title>400 Bad Request</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
<hr><center>cloudflare</center>
</body>
</html>
'
[01/28/2021 13:07:02] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/28/2021 13:07:02] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/28/2021 13:07:02] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/28/2021 13:07:02] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
```

</details>

##  Web Service Errors

The following web service errors occurred before the issue occurred:
<details>
<summary>Collapse</summary>

```
msf-ws.log does not exist.
```

</details>

##  Framework Logs

The following framework logs were recorded before the issue occurred:
<details>
<summary>Collapse</summary>

```
[01/28/2021 12:35:44] [e(0)] core: Failed to connect to the database: No database YAML file
[01/28/2021 12:35:53] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:35:53] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:35:53] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/28/2021 12:35:53] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/28/2021 12:36:49] [e(0)] core: Auxiliary failed - JSON::ParserError 809: unexpected token at '<html>
<head><title>400 Bad Request</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
<hr><center>cloudflare</center>
</body>
</html>
'
[01/28/2021 12:44:31] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:44:31] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:44:31] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/28/2021 12:44:31] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/28/2021 12:44:35] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:44:35] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:44:35] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/28/2021 12:44:35] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/28/2021 12:44:39] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:44:39] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:44:39] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/28/2021 12:44:39] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/28/2021 12:47:53] [e(0)] core: Failed to connect to the database: No database YAML file
[01/28/2021 12:47:57] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:47:57] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:47:57] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/28/2021 12:47:57] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/28/2021 12:50:27] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:50:27] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:50:27] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/28/2021 12:50:27] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/28/2021 12:55:27] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:55:27] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/28/2021 12:55:27] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/28/2021 12:55:27] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/28/2021 12:59:18] [e(0)] core: Auxiliary failed - JSON::ParserError 809: unexpected token at '<html>
<head><title>400 Bad Request</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
<hr><center>cloudflare</center>
</body>
</html>
'
[01/28/2021 13:07:02] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/28/2021 13:07:02] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/28/2021 13:07:02] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/28/2021 13:07:02] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
```

</details>

##  Web Service Logs

The following web service logs were recorded before the issue occurred:
<details>
<summary>Collapse</summary>

```
msf-ws.log does not exist.
```

</details>

##  Version/Install

The versions and install method of your Metasploit setup:
<details>
<summary>Collapse</summary>

```
Framework: 6.0.28-dev
Ruby: ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-linux-gnu]
Install Root: /usr/share/metasploit-framework
Session Type: Connected to msf. Connection type: postgresql.
Install Method: Other - Please specify
```

</details>

6. Replace these instructions and the paragraph above with the output from step 5.

**Comment by adfoster-r7** [comment]
@surfer663 Looks like there's a problem with the request; What search terms were you using?

**Comment by surfer663** [comment]
was searching for webcam i only had this problem soon after i updated kali

**Comment by adfoster-r7** [comment]
@surfer663 If this still appears, can you run the module again with `set HTTPTrace true`, and post the request/response details.

**Please delete your API key before attaching the result**

**Comment by adfoster-r7** [comment]
Looks like this is cloud flare stopping the request from completing

```
msf6 auxiliary(gather/shodan_search) > run

####################
# Request:
####################
GET /shodan/host/search?query=foo&key=XXXXXXXXXXXXXXXXXXXX&page=1 HTTP/1.1
Host: 
User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)


####################
# Response:
####################
HTTP/1.1 400 Bad Request
Server: cloudflare
Date: Fri, 29 Jan 2021 13:33:38 GMT
Content-Type: text/html
Content-Length: 155
Connection: close
CF-RAY: -

<html>
<head><title>400 Bad Request</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
<hr><center>cloudflare</center>
</body>
</html>
```

Visiting the URL directly in the browser triggers the cloudflare checks to run

**Comment by surfer663** [comment]
msf6 auxiliary(gather/shodan_search) > set HTTPTrace true
HTTPTrace => true
msf6 auxiliary(gather/shodan_search) > run

####################
# Request:
####################
GET /shodan/host/search?query=webcam&key=zScYO1R2RxkTRJ8LjkEbw29JA1K5KKJ5&page=1 HTTP/1.1                                                             
Host:                                                                      
User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)             
                                                                           
                                                                           
####################
# Response:
####################
HTTP/1.1 400 Bad Request
Server: cloudflare                                                         
Date: Fri, 29 Jan 2021 15:17:28 GMT                                        
Content-Type: text/html                                                    
Content-Length: 155                                                        
Connection: close                                                          
CF-RAY: -                                                                  
                                                                           
<html>                                                                     
<head><title>400 Bad Request</title></head>                                
<body>                                                                     
<center><h1>400 Bad Request</h1></center>                                  
<hr><center>cloudflare</center>                                            
</body>                                                                    
</html>                                                                    
                                                                           
[-] Auxiliary failed: JSON::ParserError 809: unexpected token at '<html>
<head><title>400 Bad Request</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
<hr><center>cloudflare</center>
</body>
</html>
'
[-] Call stack:
[-]   /usr/share/metasploit-framework/vendor/bundle/ruby/2.7.0/gems/json-2.5.1/lib/json/common.rb:216:in `parse'
[-]   /usr/share/metasploit-framework/vendor/bundle/ruby/2.7.0/gems/json-2.5.1/lib/json/common.rb:216:in `parse'
[-]   /usr/share/metasploit-framework/vendor/bundle/ruby/2.7.0/gems/activesupport-5.2.4.4/lib/active_support/json/decoding.rb:23:in `decode'
[-]   /usr/share/metasploit-framework/modules/auxiliary/gather/shodan_search.rb:72:in `shodan_query'
[-]   /usr/share/metasploit-framework/modules/auxiliary/gather/shodan_search.rb:118:in `run'
[*] Auxiliary module execution completed
msf6 auxiliary(gather/shodan_search) >

**Comment by surfer663** [comment]
was there something in 1 of the kali updates that caused this

**Comment by adfoster-r7** [comment]
@surfer663 This looks shodan has introduced cloudfront to their API - which is causing issues now

**Comment by surfer663** [comment]
i reset my apikey thanks.... is there nothing that can be done or is shodan going to sort this issue

**Comment by adfoster-r7** [comment]
@surfer663 Confirming with Shodan might make sense, I think it's out of our control - there's no mention of it in their API documentation from what I can tell

**Comment by surfer663** [comment]
which info would you recommend i send to shodan
many thanks

**Comment by achillean** [comment]
I believe the problem is that the browser user agent is triggering a check by Cloudflare. If you use the default user-agent for the library then it should work fine.

**Comment by adfoster-r7** [comment]
@achillean Thanks for the pointer; Looks like the vhost was being dropped after a recent pull request https://github.com/rapid7/metasploit-framework/pull/14609

We'll have to revert / put up a fix for that, sorry for the inconvenience @surfer663 :+1:

**Comment by surfer663** [comment]
> @achillean Thanks for the pointer; Looks like the vhost was being dropped after a recent pull request #14609
> 
> We'll have to revert / put up a fix for that, sorry for the inconvenience @surfer663 +1

hi achillean - is there a problem within metasploit

**Comment by surfer663** [comment]
sorry guys im a real newbie im tring to understand this but getting confused is it possible to explain to me how i can apply this fix in newbie language
many thanks

**Comment by adfoster-r7** [comment]
@surfer663 These were the required code changes:
https://github.com/rapid7/metasploit-framework/pull/14693/files

The issue will be fixed and available in Metasploit 6.0.29

**Comment by surfer663** [comment]
> @surfer663 These were the required code changes:
> https://github.com/rapid7/metasploit-framework/pull/14693/files
> 
> The issue will be fixed and available in Metasploit 6.0.29

thank you kindly
